platform-sdk/docs/proposals/TSS-Ledger-Id/TSS-Ledger-Id.md

```diff
@@ -0,0 +1,826 @@
+knows.
+
+![TSS re-keying next consensus roster](./Groth21-re-keying-shares-of-shares.drawio.svg)
+
+This process of re-keying the next consensus roster during an address book change takes an asynchronous amount of
+time through multiple rounds of consensus to complete. This requires saving incremental progress in the state to ensure
+that the process can resume from any point if a node or the network restarts. Switching to the next consensus roster
+cannot happen until that roster has enough nodes able to recover a threshold number of shares so that an aggregation of
+node signatures can generate the ledger signature.
+
+##### Elliptic Curve Decisions
+
+Our first implementation of Groth21 will use the ALT_BN128 elliptic curve which is in use and verifiable by EVMs.  
+If and when the Ethereum ecosystem adopts BLS12_381, we will likely switch over to that more secure curve.
+
+##### New Elliptic Curve Node Keys
+
+Each node will need a new long-term EC key pair in addition to the existing RSA key pair. The EC key pair will be
+used in the Groth21 algorithm to generate and decrypt `TSS-Messages`. These new EC Keys will not be used for
+signing messages, only for generating the shares. It is the share keys that are used to sign messages. The public
+keys of these long-term node specific EC keys must be in the address book.
+
+##### Groth21 Drawbacks
+
+The Groth21 algorithm is not able to model arbitrary precision proportions of weight assigned to nodes in the
+network. For example if each node in a 4 node network received 1 share, then every share has a weight of 1/4.
+If there are a total of N shares, then the distribution of weight can only be modeled in increments of (1/N) * total
+weight. The cost of re-keying the network in an address book change is quadratic in the number of shares. (See the
+picture of TssMessages above.) This forces us to pick a number of total shares with a max value in the thousands.
+This modeling of weight uses small integer precision.
+
+##### Calculating Number of Shares from Weight
+
+Given that we have a low resolution for modeling weight, the number of shares assigned to each node will be
+calculated as follows:
+
+* Let `M` be the max number of shares that any node can have.
+* Let `maxweight` be the max weight assigned to any node.
+* Let `weight(node)` be the weight assigned to the node.
+* Let `numShares(node)` be the number of shares assigned to the node.
+* then `numShares(node) = ceiling(M * weight(node) / maxweight)` where the operations are floating point.
+
+##### Threshold for recovery of Signatures and Ledger Id.
+
+Our public claim is that we can tolerate up to 1/3 of the stake assigned to malicious nodes. Our
+threshold for recovery of ledger signatures and the ledger id must be strictly greater than the following values:
+
+1. `floor(TS/3)` where `TS` is the total number of shares.
+2. `M` where `M` is the max number of shares assigned to any node.
```

Comment:
i'm confused, M is not used in the calculation below. Should it be?

Reply:
Also, it's not clear to me how the math shows that if 1/3 nodes by stake are bad actors, that it is impossible for them to recover the secret. Could you add a clear explanation for that here?

Reply:
After a conversation with Leemon, `M` cannot be a lower bound to the threshold due to support for 1 and 2 node networks.  Networks with 1 or 2 nodes must trust that the nodes are honest because one of the nodes will be able to recover the ledger private key if they were dishonest.

Reply:
I've removed the 2nd criteria.